leftwm/src/bin/leftwm-check.rs

```diff
@@ -88,13 +92,22 @@ pub fn load_from_file(fspath: Option<&str>, verbose: bool) -> Result<Config> {
             dbg!(&contents);
         }
 
+        #[cfg(feature = "toml-config")]
         let config = toml::from_str(&contents)?;
+        #[cfg(feature = "ron-config")]
+        let config = ron::from_str(&contents)?;
         Ok(config)
     } else {
         let config = Config::default();
+        #[cfg(feature = "toml-config")]
```

Comment:
A general suggestion: Would it make sense to abstract the way how we are handling config files? Because that would probably help to switch between config files. On the other hand, this would probably create some unnecessary code if we are switching only this time to another config-file type but it would definetely help, to avoid this `#[cfg(feature = "toml-config"]` for so many lines.

Reply:
Well, basically that arlready happend to some extent, with the librarification of `leftwm-core`, so it appears more or less the only places I had to touch where the actual interaction with the files, plus a couple of related tests.
I was surprised, how easy it actually was.

But if you have some ideas how to further generalise things, I am open to suggestions.

One thing that currently bugs me is that those feature-gates seem to fail, when building with `--all-targets --all-features`, which the CI does. Hence the failed checks on github.

Reply:
yeah, might want to change that CI now that we are adding mutually exclusive features. Could also add (incase both are added at the same time) an alternative path  ```#[cfg(all(feature = "toml-config", feature = "ron-config"))]
compile_error!("feature \"toml-config\" and feature \"ron-config\" cannot be enabled at the same time");``` or default to one over the other

Reply:
> But if you have some ideas how to further generalise things, I am open to suggestions.

I think that I could try to implement this. Like creating a trait for specifique config files.

## In general
I could imagine to have a directory in `leftwm-core` to have it like this:

`config/mod.rs`
```rust

pub fn get_config() {
    #[cfg(feature = "toml")]
    TomlConfigParser::new().get_config()

    #[cfg(feature = "yaml")]
    YamlConfigParser::new().get_config()
}

pub struct Config {
    /// we'd need to discuss which values are necessary for leftwm
}

pub trait ConfigParser {

    /// parses its config file. For toml it would be `~/.config/leftwm/config.toml` or for yaml `~/.config/leftwm/config.yaml` etc.
    /// Each parser would be a separate struct (see below)
    Config get_config(&self) -> Config;

    /// ...
}
````

`config/filetype/toml.rs`
```rust
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct TomlConfigParser {
    /// the toml fields
}

impl ConfigParser for TomlConfigParser {
    /// ...
}
```

`config/filetype/yaml.rs`
```rust
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct YamlConfigParser {
    /// the yaml fields
}

impl ConfigParser for YamlConfigParser {
    /// ...
}
```

Reply:
Awesome idea.

That would also be a great place to do some tweaking on the parser output as far as possible, like adding some basic doc comments and structuring to the output config file. And in case of toml try to make the parser use the compact syntax for keybinds etc.
TBH I have no idea, if the respectable serialisers have such formatting capabilities, just throwing ideas.

Reply:
Just stumbled over this possibly helpful fella: `toml::ser::tables_last` in the docs of the toml crate.

@TornaxO7 if you like to get into this, I guess a PR to this branch on my fork would probably be the easiest way to go, right?

Reply:
> @TornaxO7 if you like to get into this, I guess a PR to this branch on my fork would probably be the easiest way to go, right?

Theoretically yes, but I don't know how long it takes for me to do that since the university kicks in at the moment.
I'd say that you can merge that first and I'll create a PR afterwards.

Reply:
Ok then I guess I will work to solve the CI issues and call this review-ready.

Reply:
I'd close this because we already discussed on discord that this feature is not needed :)